GH-462: Add exponential backoff to host watcher

sergioifg94 opened the pull request:
Closes #462 


## Description of Changes

Implement exponential back off when the host watcher fails to resolve a host

## Release and Testing

1. Run local-setup using a CRC cluster (make local-setup USE_CRC=true)
2. Create the echo sample service
3. Stop the crc cluster
4. Look at the logs (Note this will happen for every DNSRecord pointing at that same cluster, so the more there are, the more logs there will be)
    * The logs will not be so polluted as the host watcher waits before retrying with exponential backoff


## Dependencies
<!-- If this PR relies on other work or if it should be merged after a certain date, detail that here. Remove if not needed-->

- [ ] Perequisites that must be completed before merging this pull request

maleck13 commented on pkg/_internal/util/utils.go:
```diff
@@ -0,0 +1,17 @@
+package util
```
perhaps we could move this to a `numbers` or `math` package rather than util

maleck13 commented on pkg/dns/host_watcher.go:
```diff
@@ -105,9 +111,15 @@ func (w *RecordWatcher) watch(ctx context.Context) {
 				w.logger.Error(err, "Failed to lookup IP address")
+				sleepTime := w.errInterval
+				w.errInterval = w.errInterval * time.Duration(2)
+				time.Sleep(util.Max(sleepTime, maxErrorInterval))
```
so this is always going to be a default of 5 mins until sleepTime reaches something > than 5 mins? The `maxErrorInterval` is an odd name for it. Maybe I have missed something but it actually seems like the `minErrorInterval`?

maleck13 replied:
Sudo code change below that might improve readability 
`
var minErrorInterval time.Duration = time.Seconds * 30				
time.Sleep(util.Max(w.errInterval, minErrorInterval))
w.errInterval = minErrorInterval * time.Duration(2)
`

sergioifg94 replied:
Thanks @maleck13  I made a mistake there. What I should have done was `Min` instead of `Max`, to limit how long the interval is, as it grows exponentially so it wouldn't take to long to become an unrealistically long interval.

Another thing I noticed I have wrong there is that the `errInterval` is not growing exponentially but linearly, it should be calculated to the power of 2, not multiplied by 2

maleck13 reviewed (commented):
left some comments about the impl

sergioifg94 commented:
Just pushed some changes addressing the issues. So the error interval at the moment grows exponentially starting at 2 seconds: 2, 4, 8, 16... but stops growing once it reaches 5 minutes

Also moved the new utility functions to a `math` package, and added some unit tests for them
